mkosi/__init__.py

```diff
@@ -3721,20 +3722,63 @@ def make_tar(args: CommandLineArguments, root: str, do_run_build_script: bool, f
     if for_cache:
         return None
 
-    tar_root_dir = f"{root}/usr" if args.usr_only else root
+    root_dir = f"{root}/usr" if args.usr_only else root
 
     with complete_step("Creating archive"):
         f: BinaryIO = cast(BinaryIO, tempfile.NamedTemporaryFile(dir=os.path.dirname(args.output), prefix=".mkosi-"))
         # OpenMandriva defaults to bsdtar(libarchive) which uses POSIX argument list so let's keep a separate list
         if shutil.which("bsdtar") and args.distribution == Distribution.openmandriva:
-            _tar_cmd = ["bsdtar", "-C", tar_root_dir, "-c", "-J", "--xattrs", "-f", "-", "."]
+            _cmd = ["bsdtar", "-C", root_dir, "-c", "-J", "--xattrs", "-f", "-", "."]
         else:
-            _tar_cmd = ["tar", "-C", tar_root_dir, "-c", "-J", "--xattrs", "--xattrs-include=*"]
+            _cmd = ["tar", "-C", root_dir, "-c", "-J", "--xattrs", "--xattrs-include=*"]
             if args.tar_strip_selinux_context:
-                _tar_cmd.append("--xattrs-exclude=security.selinux")
-            _tar_cmd.append(".")
+                _cmd.append("--xattrs-exclude=security.selinux")
+            _cmd.append(".")
 
-        run(_tar_cmd, env={"XZ_OPT": "-T0"}, stdout=f)
+        run(_cmd, env={"XZ_OPT": "-T0"}, stdout=f)
+
+    return f
+
+
+def file_list(root: str) -> Generator[str, None, None]:
+    root = root.rstrip("/") + "/"  # make sure the path ends in exactly one '/'
+    prefix_len = len(root)
+    queue = collections.deque([root])
+
+    while queue:
+        for entry in os.scandir(queue.pop()):
+            yield entry.path[prefix_len:]
+            if entry.is_dir(follow_symlinks=False):
+                queue.append(entry.path)
+
+
+def make_cpio(args: CommandLineArguments, root: str, do_run_build_script: bool, for_cache: bool) -> Optional[BinaryIO]:
+    if do_run_build_script:
+        return None
+    if args.output_format != OutputFormat.cpio:
+        return None
+    if for_cache:
+        return None
+
+    root_dir = f"{root}/usr" if args.usr_only else root
+
+    with complete_step("Creating archive"):
+        f: BinaryIO = cast(BinaryIO, tempfile.NamedTemporaryFile(dir=os.path.dirname(args.output), prefix=".mkosi-"))
+
+        with delay_interrupt():
+            files = file_list(root_dir)
+
+            with subprocess.Popen(
+                ["cpio", "-o", "--reproducible", "--null", "-H", "newc", "--quiet", "-D", root_dir],
+                stdin=subprocess.PIPE,
+                stdout=subprocess.PIPE,
+            ) as cpio:
+                with subprocess.Popen(
+                    ["pxz", "--check=crc32", "--lzma2=dict=1MiB", "-T0"], stdin=cpio.stdout, stdout=f
```

Comment:
share the infra from xz_binary() here to use pxz only if installed, and use plain xz otherwise

Reply:
That chunk is replaced later on with something that does the check…